File: pi_integration/src/pi_integration_omp.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>

int main (int argc, char *argv[])
{
  int i,num_steps;
  double x, sum, step, pi;
  double t_start, t_end;
  
  num_steps=500000000;


  x=0;
  sum = 0.0;
  step = 1.0/(double) num_steps;
  t_start = omp_get_wtime();
#pragma omp parallel for reduction(+:sum) default(none) private(i,x) firstprivate(step, num_steps)
  //schedule(static)
  for (i=0; i < num_steps; i++)
    {
      x = (i+0.5)*step;
      sum = sum + 4.0/(1.0+x*x);
    }
  
  pi = step * sum;
  t_end = omp_get_wtime();
  printf("Value of pi = %g\n",pi);
  printf("Expended wall clock time = %.20f\n", t_end - t_start);
  printf("num proc = %d\n", omp_get_num_procs());
  return EXIT_SUCCESS;
    
}

File: pi_integration/src/pi_integration_omp.f90
PROGRAM pi_integration_omp
  USE omp_lib
  IMPLICIT NONE
  INTEGER :: i
  REAL(kind=8) :: x, sum, step, pi
  REAL(kind=8) :: t_start, t_end

  INTEGER(kind=8), PARAMETER :: num_steps = 500000000

  sum=0.0
  step = 1.0/num_steps

  t_start = omp_get_wtime()
  !$OMP PARALLEL DO DEFAULT(none)        &
  !$OMP SHARED(pi,step)                  &
  !$OMP PRIVATE(x)                       &
  !$OMP SCHEDULE(static) REDUCTION(+:sum)
  DO i=0,num_steps
     x = (i+0.5)*step
     sum = sum + 4.0/(1.0+x*x)
  END DO
  !$OMP END PARALLEL DO

  pi = step * sum

  t_end = omp_get_wtime()

  WRITE(*,*) "Value of pi = ",pi
  WRITE(*,*) "Expended wall clock time = ",t_end-t_start
  
END PROGRAM pi_integration_omp
